rand_distr/tests/cdf.rs

```diff
@@ -195,3 +374,81 @@ fn binomial() {
         });
     }
 }
+
+#[test]
+fn geometric() {
+    fn cdf(k: i64, p: f64) -> f64 {
+        if k < 0 {
+            0.0
+        } else {
+            1.0 - (1.0 - p).powi(1 + k as i32)
+        }
+    }
+
+    let parameters = [0.3, 0.5, 0.7, 0.0000001, 0.9999];
+
+    for (seed, p) in parameters.into_iter().enumerate() {
+        let dist = rand_distr::Geometric::new(p).unwrap();
+        test_discrete(seed as u64, dist, |k| cdf(k, p));
+    }
+}
+
+#[test]
+fn hypergeometric() {
+    fn cdf(x: i64, n: u64, k: u64, n_: u64) -> f64 {
+        let min = if n_ + k > n { n_ + k - n } else { 0 };
+        let max = k.min(n_);
+        if x < min as i64 {
+            return 0.0;
+        } else if x >= max as i64 {
+            return 1.0;
+        }
+
+        (min..x as u64 + 1).fold(0.0, |acc, k_| {
+            acc + (ln_binomial(k, k_) + ln_binomial(n - k, n_ - k_) - ln_binomial(n, n_)).exp()
+        })
+    }
+
+    let parameters = [
+        (15, 13, 10),
+        (25, 15, 5),
+        (60, 10, 7),
+        (70, 20, 50),
+        (100, 50, 10),
+        // (100, 50, 49), // Fail case
+    ];
+
+    for (seed, (n, k, n_)) in parameters.into_iter().enumerate() {
+        let dist = rand_distr::Hypergeometric::new(n, k, n_).unwrap();
+        test_discrete(seed as u64, dist, |x| cdf(x, n, k, n_));
+    }
+}
+
+#[test]
+fn poisson() {
+    fn cdf(k: i64, lambda: f64) -> f64 {
+        if k < 0 || lambda <= 0.0 {
+            return 0.0;
+        }
+
+        1.0 - lambda.inc_gamma(k as f64 + 1.0)
+    }
+
+    let parameters = [
+        0.1_f32, 1.0, 7.5,
+        // 1.844E+19,  // fail case
```

Comment:
This case fails because your cdf does not work for values bigger than `1e8`. The problem is in the inc_gamma though, it just return 1.0 when `lambda > 1e8`. Maybe you can find a different implementation.

Reply:
It might not be related to the implementation of `lgamma`, but rather that the probability is too small to be stored in `f64`. In fact, `cdf(1e8 as i64, 1e8)` can output the result `0.5000265872649796`, but `cdf((1e8 - 1e5) as i64, 1e8)` rounds to `0`. I tried using `f128`, but encountered linker-related issues, and even if it worked, it wouldn't be usable in the stable version.

I will also check other implementations of the CDF to see if there is a possibility of underflow.

Reply:
cdf(2e9, 1e9) should be 1.0 but it is 0.0

Reply:
In the latest commit, I copied the [`gamma_lr`](https://docs.rs/statrs/latest/src/statrs/function/gamma.rs.html#260-347) implementation from `statrs` and used it to calculate the CDF:

```rust
println!("{}", cdf(1e9 as i64, 1e9)); // 0.5000077942086617
println!("{}", cdf(2e9 as i64, 1e9)); // 1
println!("{}", cdf((1e9 - 1e4) as i64, 1e9)); // 0.3759229845139799
```

For reference, here is the calculation in R:

```R
> ppois(1e9, 1e9)
[1] 0.5000084
> ppois(2e9, 1e9)
[1] 1
> ppois(1e9-1e4, 1e9)
[1] 0.3759226
```